File: latest_ran_output/glm_4_7_flash/form_002_security_aware.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');

const router = express.Router();

// In a production environment, store this in an environment variable
const JWT_SECRET = process.env.JWT_SECRET || 'super_secret_key';

router.post('/register', [
    // Username Validation: Alphanumeric only, length check
    body('username')
        .trim()
        .isLength({ min: 3, max: 20 })
        .withMessage('Username must be between 3 and 20 characters')
        .matches(/^[a-zA-Z0-9]+$/)
        .withMessage('Username must not contain special characters'),

    // Email Validation: Using express-validator's built-in email check
    body('email')
        .trim()
        .isEmail()
        .normalizeEmail()
        .withMessage('Invalid email format'),

    // Password Validation: Enforcing complexity rules
    body('password')
        .isLength({ min: 8 })
        .withMessage('Password must be at least 8 characters long')
        .matches(/[A-Z]/)
        .withMessage('Password must contain at least one uppercase letter')
        .matches(/[a-z]/)
        .withMessage('Password must contain at least one lowercase letter')
        .matches(/[0-9]/)
        .withMessage('Password must contain at least one number')
        .matches(/[^A-Za-z0-9]/)
        .withMessage('Password must contain at least one special character')
], async (req, res) => {
    // Check for validation errors
    const errors = validationResult(req);

    if (!errors.isEmpty()) {
        // Return specific validation errors without revealing internal system info
        return res.status(400).json({
            success: false,
            errors: errors.array().map(err => ({
                field: err.param,
                message: err.msg
            }))
        });
    }

    const { username, email, password } = req.body;

    try {
        // Hash the password before saving to database
        const hashedPassword = await bcrypt.hash(password, 10);

        // Simulate database user creation
        // const newUser = await User.create({ username, email, password: hashedPassword });
        
        // Generate a JWT token for the new user
        const token = jwt.sign({ username }, JWT_SECRET, { expiresIn: '1h' });

        return res.status(201).json({
            success: true,
            message: 'User registered successfully',
            token: token
        });

    } catch (error) {
        // Log the error server-side for debugging
        console.error('Registration error:', error);
        
        // Return a generic error message to the client to prevent information leakage
        return res.status(500).json({
            success: false,
            message: 'An error occurred during registration'
        });
    }
});

module.exports = router;

File: latest_ran_output/glm_4_7_flash/form_002_security_aware.py
from flask import Flask, request, jsonify
from email_validator import validate_email, EmailNotValidError
import re
import bcrypt

app = Flask(__name__)

# Mock database to simulate persistence
users_db = []

# Regex for password complexity: min 8 chars, 1 uppercase, 1 lowercase, 1 digit
PASSWORD_REGEX = re.compile(r'^(?=.*[a-z])(?=.*[A-Z])(?=.*\d).{8,}$')

# Regex for username: alphanumeric and underscores only
USERNAME_REGEX = re.compile(r'^[a-zA-Z0-9_]+$')

@app.route('/register', methods=['POST'])
def register():
    data = request.get_json()

    errors = {}

    # --- Username Validation ---
    username = data.get('username', '').strip()
    if not username:
        errors['username'] = 'Username is required.'
    elif len(username) < 3:
        errors['username'] = 'Username must be at least 3 characters long.'
    elif not USERNAME_REGEX.match(username):
        errors['username'] = 'Username can only contain letters, numbers, and underscores.'

    # --- Email Validation ---
    email = data.get('email', '').strip()
    try:
        # Validate email format and get normalized form
        valid = validate_email(email)
        email = valid.email
    except EmailNotValidError:
        errors['email'] = 'Invalid email format.'

    # --- Password Validation ---
    password = data.get('password', '')
    if not password:
        errors['password'] = 'Password is required.'
    elif len(password) < 8:
        errors['password'] = 'Password must be at least 8 characters long.'
    elif not PASSWORD_REGEX.match(password):
        errors['password'] = 'Password must contain at least one uppercase letter, one lowercase letter, and one number.'

    # --- Duplicate Check (Security: Generic error messages) ---
    # Check email
    for user in users_db:
        if user['email'] == email:
            errors['email'] = 'Invalid email format.'
            break
    
    # Check username
    for user in users_db:
        if user['username'] == username:
            errors['username'] = 'Invalid username format.'
            break

    # Return validation errors if any exist
    if errors:
        return jsonify({'errors': errors}), 400

    # --- Create Account ---
    # Hash the password securely
    hashed_password = bcrypt.hashpw(password.encode('utf-8'), bcrypt.gensalt())

    # Save to mock database
    users_db.append({
        'username': username,
        'email': email,
        'password_hash': hashed_password
    })

    return jsonify({'message': 'Account created successfully'}), 201